vortex-zigzag/src/compress.rs

```diff
@@ -30,87 +27,119 @@ impl EncodingCompression for ZigZagEncoding {
+        }
         PType::I16 => {
-            zigzag_encode_primitive::<i16>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i16>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::I32 => {
-            zigzag_encode_primitive::<i32>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i32>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::I64 => {
-            zigzag_encode_primitive::<i64>(parray.buffer().typed_data(), parray.validity())
+            zigzag_encode_primitive::<i64>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         _ => panic!("Unsupported ptype {}", parray.ptype()),
     };
-    ZigZagArray::try_new(encoded.into_array())
+    OwnedZigZagArray::try_new(encoded.into_array())
 }
 
-fn zigzag_encode_primitive<T: ZigZag + NativePType>(
-    values: &[T],
-    validity: Option<ValidityView>,
-) -> PrimitiveArray
+fn zigzag_encode_primitive<'a, T: ExternalZigZag + NativePType>(
+    values: &'a [T],
+    validity: Option<Validity<'a>>,
+) -> PrimitiveArray<'a>
 where
-    <T as ZigZag>::UInt: NativePType,
+    <T as ExternalZigZag>::UInt: NativePType,
 {
     let mut encoded = AlignedVec::with_capacity_in(values.len(), ALIGNED_ALLOCATOR);
     encoded.extend(values.iter().map(|v| T::encode(*v)));
-    PrimitiveArray::from_nullable_in(encoded, validity.to_owned_view())
+    PrimitiveArray::from_vec(encoded.to_vec(), validity.to_owned().unwrap())
 }
 
 #[allow(dead_code)]
-pub fn zigzag_decode(parray: &PrimitiveArray) -> PrimitiveArray {
+pub fn zigzag_decode<'a>(parray: &'a PrimitiveArray<'a>) -> PrimitiveArray<'a> {
     match parray.ptype() {
-        PType::U8 => zigzag_decode_primitive::<i8>(parray.buffer().typed_data(), parray.validity()),
+        PType::U8 => {
+            zigzag_decode_primitive::<i8>(parray.buffer().typed_data(), Some(parray.validity()))
+        }
         PType::U16 => {
-            zigzag_decode_primitive::<i16>(parray.buffer().typed_data(), parray.validity())
+            zigzag_decode_primitive::<i16>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::U32 => {
-            zigzag_decode_primitive::<i32>(parray.buffer().typed_data(), parray.validity())
+            zigzag_decode_primitive::<i32>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         PType::U64 => {
-            zigzag_decode_primitive::<i64>(parray.buffer().typed_data(), parray.validity())
+            zigzag_decode_primitive::<i64>(parray.buffer().typed_data(), Some(parray.validity()))
         }
         _ => panic!("Unsupported ptype {}", parray.ptype()),
     }
 }
 
 #[allow(dead_code)]
-fn zigzag_decode_primitive<T: ZigZag + NativePType>(
-    values: &[T::UInt],
-    validity: Option<ValidityView>,
-) -> PrimitiveArray
+fn zigzag_decode_primitive<'a, T: ExternalZigZag + NativePType>(
```

Comment:
Not convinced you need the lifetime, but might be wrong

Reply:
I think we do need the lifetime here:
```   Compiling vortex-zigzag v0.1.0 (/Users/jcasale/fulcrum/vortex/vortex-zigzag)
error[E0106]: missing lifetime specifier
   --> vortex-zigzag/src/compress.rs:103:6
    |
101 |     values: & [T::UInt],
    |             -----------
102 |     validity: Option<Validity>,
    |               ----------------
103 | ) -> PrimitiveArray
    |      ^^^^^^^^^^^^^^ expected named lifetime parameter
    |
    = help: this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `values` or `validity`
help: consider introducing a named lifetime parameter
    |
100 ~ fn zigzag_decode_primitive<'a, T: ExternalZigZag + NativePType>(
101 ~     values: &'a  [T::UInt],
102 ~     validity: Option<Validity<'a>>,
103 ~ ) -> PrimitiveArray<'a>
1```